auth0/v3/authentication/token_verifier.py

```diff
@@ -0,0 +1,225 @@
+import json
+import time
+
+import jwt
+import requests
+
+from auth0.v3.exceptions import TokenValidationError
+
+
+class SignatureVerifier():
+    DISABLE_JWT_CHECKS = {
+        "verify_signature": True,
+        "verify_exp": False,
+        "verify_nbf": False,
+        "verify_iat": False,
+        "verify_aud": False,
+        "verify_iss": False,
+        "require_exp": False,
+        "require_iat": False,
+        "require_nbf": False,
+    }
+
+    def __init__(self, algorithm):
+        if not algorithm or type(algorithm) != str:
+            raise ValueError("algorithm must be specified.")
+        self._algorithm = algorithm
+
+    def _fetch_key(self, key_id=None):
+        raise NotImplementedError
+
+    def verify_signature(self, token):
+        try:
+            header = jwt.get_unverified_header(token)
+        except jwt.exceptions.DecodeError:
+            raise TokenValidationError("ID token could not be decoded.")
+
+        alg = header.get('alg', None)
+        if alg != self._algorithm:
+            raise TokenValidationError('Signature algorithm of "{}" is not supported. Expected the ID token to be signed with "{}"'.format(alg, self._algorithm))
+
+        kid = header.get('kid', None)
+        secret_or_certificate = self._fetch_key(key_id=kid)
+
+        try:
+            decoded = jwt.decode(jwt=token, key=secret_or_certificate, algorithms=[self._algorithm], options=self.DISABLE_JWT_CHECKS)
+        except jwt.exceptions.InvalidSignatureError:
+            raise TokenValidationError("Invalid token signature.")
+        return decoded
+
+
+class SymmetricSignatureVerifier(SignatureVerifier):
+
+    def __init__(self, shared_secret, algorithm="HS256"):
+        SignatureVerifier.__init__(self, algorithm)
+        self._shared_secret = shared_secret
+
+    def _fetch_key(self, key_id=None):
+        return self._shared_secret
+
+
+class AsymmetricSignatureVerifier(SignatureVerifier):
+
+    def __init__(self, jwks_url, algorithm="RS256", _jwks_fetcher=None):
+        SignatureVerifier.__init__(self, algorithm)
+        self._fetcher = _jwks_fetcher or JwksFetcher(jwks_url)
+
+    def _fetch_key(self, key_id=None):
+        return self._fetcher.get_key(key_id)
+
+
+class JwksFetcher():
+    CACHE_TTL = 600  # 10 min cache lifetime
+
+    def __init__(self, jwks_url, cache_ttl=CACHE_TTL):
+        self._jwks_url = jwks_url
+        self._init_cache(cache_ttl)
+        return
+
+    def _init_cache(self, cache_ttl):
+        self._cache_value = {}
+        self._cache_date = 0
+        self._cache_ttl = cache_ttl
+        self._cache_is_fresh = False
+
+    def _fetch_jwks(self, force=False):
```

Comment:
the logic here is self-explanatory, in short:
if not asked to forcefully make a request or the token has not expired, return value from the cache. In any other case, make the request and update the cache.